src/type-factor.c

```diff
@@ -74,3 +74,287 @@ static SEXP levels_union(SEXP x, SEXP y) {
+  R_len_t size = vec_size(levels);
+  const SEXP* p_levels = STRING_PTR_RO(levels);
+
+  // There would only ever be 1 `NA` level
+  int na_loc;
+  bool any_na = false;
+
+  for (R_len_t i = 0; i < size; ++i) {
+    if (p_levels[i] != NA_STRING) {
+      continue;
+    }
+
+    any_na = true;
+    na_loc = (i + 1) * -1;
+    break;
+  }
+
+  // Remove `NA` level if required
+  if (any_na) {
+    return vec_slice(levels, r_int(na_loc));
+  } else {
+    return levels;
+  }
+}
+
+
+static SEXP fct_as_factor_impl(SEXP x, SEXP x_levels, SEXP to_levels, bool* lossy, bool ordered);
+
+// [[ include("vctrs.h") ]]
+SEXP fct_as_factor(SEXP x,
+                   SEXP to,
+                   bool* lossy,
+                   struct vctrs_arg* x_arg,
+                   struct vctrs_arg* to_arg) {
+
+  SEXP x_levels = Rf_getAttrib(x, R_LevelsSymbol);
+  SEXP to_levels = Rf_getAttrib(to, R_LevelsSymbol);
+
+  if (TYPEOF(x_levels) != STRSXP) {
+    stop_corrupt_factor_levels(x, x_arg);
+  }
+
+  if (TYPEOF(to_levels) != STRSXP) {
+    stop_corrupt_factor_levels(to, to_arg);
+  }
+
+  return fct_as_factor_impl(x, x_levels, to_levels, lossy, false);
+}
+
+// [[ include("vctrs.h") ]]
+SEXP ord_as_ordered(SEXP x,
+                    SEXP to,
+                    bool* lossy,
+                    struct vctrs_arg* x_arg,
+                    struct vctrs_arg* to_arg) {
+
+  SEXP x_levels = Rf_getAttrib(x, R_LevelsSymbol);
+  SEXP to_levels = Rf_getAttrib(to, R_LevelsSymbol);
+
+  if (TYPEOF(x_levels) != STRSXP) {
+    stop_corrupt_ordered_levels(x, x_arg);
+  }
+
+  if (TYPEOF(to_levels) != STRSXP) {
+    stop_corrupt_ordered_levels(to, to_arg);
+  }
+
+  return fct_as_factor_impl(x, x_levels, to_levels, lossy, true);
+}
+
+static SEXP fct_as_factor_impl(SEXP x, SEXP x_levels, SEXP to_levels, bool* lossy, bool ordered) {
+  // Early exit if levels are identical
+  if (x_levels == to_levels) {
+    return x;
+  }
+
+  R_len_t x_levels_size = vec_size(x_levels);
+  R_len_t to_levels_size = vec_size(to_levels);
+
+  // Early exit if `to` has no levels. In this case it is being used as
+  // a template
+  if (to_levels_size == 0) {
+    return x;
+  }
+
+  // Always lossy if there are more levels in `x` than in `to`
+  if (x_levels_size > to_levels_size) {
+    *lossy = true;
+    return R_NilValue;
+  }
+
+  R_len_t x_size = vec_size(x);
+
+  const SEXP* p_x_levels = STRING_PTR_RO(x_levels);
+  const SEXP* p_to_levels = STRING_PTR_RO(to_levels);
+
+  bool is_contiguous_subset = true;
+
+  for (R_len_t i = 0; i < x_levels_size; ++i) {
+    if (p_x_levels[i] != p_to_levels[i]) {
+      is_contiguous_subset = false;
+      break;
+    }
+  }
+
+  const int* p_x = INTEGER_RO(x);
+
+  SEXP out = PROTECT(Rf_allocVector(INTSXP, x_size));
+  int* p_out = INTEGER(out);
+
+  if (ordered) {
+    new_ordered(out, to_levels);
+  } else {
+    new_factor(out, to_levels);
+  }
+
+  // No recode required
+  if (is_contiguous_subset) {
+    memcpy(p_out, p_x, x_size * sizeof(int));
```

Comment:
`Rf_shallow_duplicate()` instead? Might help avoid a copy thanks to ALTREP wrappers.

Reply:
That worked nicely!